File: app/helpers/drive.py
from apiclient import discovery
from ..config import *

def add_and_share_drive_folder(http_auth, contractor, email_created):
    # Wait until e-mail has been created for contractor
    email_created.wait()

    # TODO: Take me out for production!!!
    # return {"text": "Set up new folder in Google Drive", "status": "success"}

    try:
        drive = discovery.build("drive", "v2", http_auth)

        folder_id = create_folder(drive, contractor.get_full_name())
        add_files(drive, folder_id, contractor.is_resident)
        share_folder(drive, folder_id, contractor.get_email())

        return {"text": "Set up new folder in Google Drive", "status": "success"}
    except:
        return {"text": "Problem with setting up Google Drive", "status": "failure"}


def create_folder(drive, name):
    file_metadata = {
        "title": name,
        "parents": [{"id": HR_CONTRACTOR_FOLDER}],
        "mimeType": "application/vnd.google-apps.folder"}
    file = drive.files().insert(body=file_metadata, fields="id").execute()

    return file.get("id")


def add_files(drive, folder_id, is_resident):
    files_to_share = [{"id": BACKGROUND_CHECK, "title": "Background Check Authorization"},
                      {"id": DIRECT_DEPOSIT, "title": "Direct Deposit Form"},
                      {"id": MSSA, "title": "MSSA"}]
    if is_resident:
        files_to_share.append({"id": W9, "title": "Form W-9"})
    else:
        files_to_share.append({"id": W8, "title": "Form W-8BEN"})

    hr_forms_folder = HR_FORMS_FOLDER

    for file_to_share in files_to_share:
        copy_body = {"title": file_to_share["title"]}
        copy_file = drive.files().copy(fileId=file_to_share["id"], body=copy_body).execute()
        copy_file_id = copy_file.get("id")
        drive.files().update(fileId=copy_file_id, addParents=folder_id, removeParents=hr_forms_folder).execute()


def share_folder(drive, folder_id, email):
    user_permission = {
        "type": "user",
        "role": "writer",
        "emailAddress": email,
        "value": email}
    shared = drive.permissions().insert(
        fileId=folder_id,
        body=user_permission,
        fields="id",
    ).execute()

    return shared.get("id")


def get_tasks_from_spreadsheet(http_auth):
    drive = discovery.build("drive", "v2", http_auth)
    spreadsheet = drive.files().export_media(fileId=TASKS_SPREADSHEET, mimeType="text/csv").execute()

    return str(spreadsheet).splitlines()

File: controller/drive.js
/**
 * @fileOverview Handles all calls to Google's Drive API
 */
let request = require('request-promise').defaults({simple: false});
let config = require('config');
let googleAuth = require('./google-auth');
let Bluebird = require('bluebird');

let drive = exports;

/**
 * Main function called to create the folder, add the necessary files and share the folder with the contractor
 * @param contractor
 * @param credentials Google auth credentials stored in the user's sessions
 * @returns success/failure status object
 */
drive.addAndShareDriveFolder = (contractor, credentials) => {
  return drive.createFolder(contractor, credentials)
    .then( (folderId) => {
      let toDoAfterFolderCreated = [
        drive.shareFolder(contractor, credentials, folderId),
        drive.addFile(contractor, credentials, config.get('drive.files.directDeposit'), folderId),
        drive.addFile(contractor, credentials, config.get('drive.files.bgCheck'), folderId),
      ];
      if (contractor.isResident) {
        toDoAfterFolderCreated.push(
          drive.addFile(contractor, credentials, config.get('drive.files.w9'), folderId));
      } else {
        toDoAfterFolderCreated.push(
          drive.addFile(contractor, credentials, config.get('drive.files.w8'), folderId));
      }
        return Bluebird.all(toDoAfterFolderCreated);
    })
    .then( () => {
      return {'text': 'Created and shared Drive folder', 'status': 'success'};
    })
    .catch( (err) => {
      console.log(err);
      return {'text': 'Problem creating and sharing Drive folder', 'status': 'failure'};
    })
};

/**
 * Helper function to create the folder
 * @param contractor
 * @param credentials Google auth credentials stored in the user's sessions
 * @returns created folder id
 * @throws error if no folder is created
 */
drive.createFolder = (contractor, credentials) => {
  let driveUrl = config.get('google.baseUrl') + '/drive/v3/files';

  return googleAuth.getAccessToken(credentials)
    .then((token) => {
      return request.post({
        url: driveUrl,
        qs: {access_token: token},
        json: true,
        body: {
          'name': contractor.getFullName(),
          'parents': [config.get('drive.folders.contractors')],
          'mimeType': 'application/vnd.google-apps.folder'
        }
      });
    })
    .then((response) => {
      let fileData = JSON.parse(JSON.stringify(response));
      if ('id' in fileData) {
        return fileData.id;
      } else {
        throw new Error('No id in create folder response: ' + fileData);
      }
    });
};

/**
 * Helper function to make the folder available to the contractor
 * @param contractor
 * @param credentials Google auth credentials stored in the user's sessions
 * @returns shared folder id
 * @throws error if folder is not shared
 */
drive.shareFolder = (contractor, credentials, folderId) => {
  let driveUrl = config.get('google.baseUrl') + '/drive/v3/files/'+ folderId +'/permissions';

  return googleAuth.getAccessToken(credentials)
    .then((token) => {
      return request.post({
        url: driveUrl,
        qs: {
          access_token: token,
          sendNotificationEmail: true},
        json: true,
        body: {
          'type': 'user',
          'role': 'writer',
          'emailAddress': contractor.getEmail(),
        }
      });
    })
    .then((response) => {
      let fileData = JSON.parse(JSON.stringify(response));
      if ('id' in fileData) {
        return fileData.id;
      } else {
        throw new Error('No id in share folder response: ' + fileData);
      }
    });
};

/**
 * Helper function to copy a file and then move to the copy to a folder
 * @param contractor
 * @param credentials Google auth credentials stored in the user's sessions
 * @param file An object representing the file to be copied/moved
 * @param folderId The id of the folder the file is being added to
 * @returns the id of the file that was added
 * @throws error if the file was not added
 */
drive.addFile = (contractor, credentials, file, folderId) => {
  let driveUrl = config.get('google.baseUrl') + '/drive/v3/files/'+ file.id +'/copy';

  return googleAuth.getAccessToken(credentials)
    .then((token) => {
      return request.post({
        url: driveUrl,
        qs: {access_token: token},
        json: true,
        body: {
          'name': file.name,
          'parents': [folderId],
        }
      });
    })
    .then((response) => {
      let fileData = JSON.parse(JSON.stringify(response));
      if ('id' in fileData) {
        return fileData.id;
      } else {
        throw new Error('No id in copy file response: ' + fileData);
      }
    });
};

/**
 * Helper function to retrieve a list of tasks from a spreadsheet on the Drive
 * @param credentials Google auth credentials stored in the user's sessions
 * @returns an array of CSV strings representing the tasks
 * @throws error if the tasks could not be retrieved
 */
drive.getTasksFromFile = (credentials) => {
  let driveUrl = config.get('google.baseUrl') + '/drive/v3/files/'+ config.get('drive.files.task.id') +'/export';

  return googleAuth.getAccessToken(credentials)
    .then((token) => {
      return request.get({
        url: driveUrl,
        qs: {
          access_token: token,
          mimeType: 'text/csv'
        },
        json: true
      });
    })
    .then((response) => {
      let fileData = String(response);
      if (fileData.indexOf(',') > -1) {
        return fileData.match(/[^\r\n]+/g);
      } else {
        throw new Error('Did not return CSV: ' + fileData);
      }
    });
}

